Float: losslessly convert strings to high-precision Floats

cbm755 opened the pull request:
```
Float: losslessly convert strings to high-precision Floats

Previously, the output of `sympify(<str>)` always gave precision
of 15, even if the string was much longer.  `N(<str>)` and Float
were similar.  For Float, a user could always to `Float(str, '')`
to auto select the precision.  This is good but perhaps not ideal
for `sympify(str)` (which ultimately calls `Float(str)` without
the nullstring.

The new behaviour is all of sympify, N, and Float will now default
to a minimum precision of 15 when converting from a string (thus
preserving previous behaviour).  But when given a long string
they can use additional precision automatically.

This is accomplished with a new choice of the `prec` kwarg in the
Float constructor: `prec='min15'` uses a minimum precision of 15
but can use higher for strings, where is behaves like the null
string.  This new choice is made the default.

Fixes #8821.
```

cbm755 committed:
sympify/evalf: tests for high precision from strings

Generate very accurate pi string and then try to use S and N to
create objects with more than double precision.

cbm755 committed:
Float: losslessly convert strings to high-precision Floats

Previously, the output of `sympify(<str>)` always gave precision
of 15, even if the string was much longer.  `N(<str>)` and Float
were similar.  For Float, a user could always to `Float(str, '')`
to auto select the precision.  This is good but perhaps not ideal
for `sympify(str)` (which ultimately calls `Float(str)` without
the nullstring.

The new behaviour is all of sympify, N, and Float will now default
to a minimum precision of 15 when converting from a string (thus
preserving previous behaviour).  But when given a long string
they can use additional precision automatically.

This is accomplished with a new choice of the `prec` kwarg in the
Float constructor: `prec='min15'` uses a minimum precision of 15
but can use higher for strings, where is behaves like the null
string.  This new choice is made the default.

Fixes #8821.

cbm755 committed:
test_numbers: update a failing test

Now that Float can automatically adjust precision above 15 when
given string input, there is no need to test that it cannot!
This test is being updated because it tested for a limitation rather
than the desired behaviour.  The limitation has now been removed.

cbm755 committed:
test_numbers: add Float(str) test for high precision

Test that `Float(<string>)` can use more than double precision
automatically (without passing the optional null-string second
argument).

cbm755 commented:
I'm fixing it in Float, because when I tried to fix it in `sympify` it got messy.

https://github.com/cbm755/sympy/tree/sympify_hiprec_str

Plus, I think Float should work like this too!

smichr commented:
So if I understand, you are using a string 'min15' so we can differentiate between the case of a desired 15 and the default "15 or higher if it's a string" case. Why not just set the default to None so the user has to read the docstring to know what the default is -- which is probably what they will need to do to understand 'min15', anyway? (I think 'min15' is a sane choice but using None is more pythonic.)

cbm755 committed:
Float: use "prec=None" as the default, improve docs

Drop the rather silly "min15" value, just use None.

cbm755 committed:
Float: docstring: one line summary for PEP-257

cbm755 committed:
Float: also test Float(long) high precision

cbm755 committed:
nfloat: fix tests for new Float behaviour on longs

nfloat defaults to 15 digits precision, and this test expects
`Float(<long>)` does too.  But that is no longer the case.
Change to `Float(<long>, 15)`.

Also added a missing case from the test (`big*x` as well as
the existing `x**big`).

cbm755 committed:
test_numbers: undo an unnecessary import

cbm755 commented:
Thank you @smichr that is nicer, done as you describe.

Also, fixed up documentation.

Finally, fixed the failing test.  So in total, this changes two existing tests (long int in test_numbers.py is an improvement, nfloat perhaps less obvious: my test change just maintains the status quo there).

smichr commented on sympy/core/numbers.py:
```diff
@@ -505,6 +513,19 @@ class Float(Number):
     Notes
     =====
 
+    The behaviour when using the null string argument with low-precision can
```
This seems confusing to me. Before you try to rewrite, what is the point that you are trying to highlight? That you get a low precision float if you use the ''-arg?

cbm755 replied:
1.  I suppose I was worried someone will wonder why `prec=''` is not the default.

2.  Related: when I tried making `prec=''` the default, it broke simple things like `sympify('2/2.6')` which users (and our tests!) quite reasonably expect to behave like `a = 2/2.6; b = sympify(a)`.

I could just delete this whole paragraph; I don't think I have a clear point here.  So if its confusing *and* not helpful, let's just remove?

smichr commented on sympy/core/numbers.py:
```diff
@@ -469,16 +467,26 @@ class Float(Number):
-    to force ints to Float or to enter high-precision (> 15 significant
-    digits) values:
+    Creating Floats from strings (and Python ``int`` and ``long`` types)
+    typically uses approximately 15 significant digits of precision; however
```
Perhaps delete the questionable section we discussed and rewrite this section as:

```
Creating Floats from strings (and Python ``int`` and ``long`` types)
will give a minimum precision of 15 digits, but the precision will
automatically increase to capture all digits entered.

>>> Float(1)
1.00000000000000
>>> 10**20
100000000000000000000L
>>> Float(_)
100000000000000000000.

Large *floating-point* numbers are truncated to 15 digits of precision,
however:

>>> Float(1e20)
1.00000000000000e+20

The desired number of digits can also be specified:
```

cbm755 committed:
Float: remove confusing note in doc

cbm755 committed:
Float: improve documentation

cbm755 commented:
@smichr I did similar to your suggestion.  Rebased for import conflict with master.  Last two commits are what is new.

smichr commented:
Thanks! Looks good. It's in.

cbm755 commented:
Thanks @smichr.  I added something to the release notes.
